cpp/src/strings/search/findall.cu

```diff
@@ -21,134 +21,148 @@
 
 #include <cudf/column/column.hpp>
 #include <cudf/column/column_device_view.cuh>
+#include <cudf/detail/iterator.cuh>
 #include <cudf/detail/null_mask.hpp>
 #include <cudf/detail/nvtx/ranges.hpp>
 #include <cudf/strings/detail/strings_column_factories.cuh>
-#include <cudf/strings/detail/utilities.hpp>
 #include <cudf/strings/findall.hpp>
 #include <cudf/strings/string_view.cuh>
-#include <cudf/strings/strings_column_view.hpp>
 
 #include <rmm/cuda_stream_view.hpp>
 #include <rmm/exec_policy.hpp>
 
+#include <thrust/for_each.h>
 #include <thrust/functional.h>
 #include <thrust/iterator/counting_iterator.h>
 #include <thrust/pair.h>
 #include <thrust/reduce.h>
-#include <thrust/transform.h>
 
 namespace cudf {
 namespace strings {
 namespace detail {
 using string_index_pair = thrust::pair<const char*, size_type>;
+using indices_span      = cudf::detail::device_2dspan<string_index_pair>;
 
 namespace {
 /**
- * @brief This functor handles extracting matched strings by applying the compiled regex pattern
- * and creating string_index_pairs for all the substrings.
+ * @brief This functor calls regex find on each string and creates
+ * string_index_pairs for all matching substrings.
+ *
+ * The number of output columns is dependent on the string with the most matches.
+ * For strings with fewer matches, null entries are appended into d_indices
+ * up to the maximum column count.
  */
 template <int stack_size>
 struct findall_fn {
   column_device_view const d_strings;
   reprog_device prog;
-  size_type const column_index;
-  size_type const* d_counts;
+  size_type const* d_counts;  ///< match counts for each string
+  indices_span d_indices;     ///< 2D-span: output matches added here
 
-  __device__ string_index_pair operator()(size_type idx)
+  __device__ void operator()(size_type idx)
   {
-    if (d_strings.is_null(idx) || (column_index >= d_counts[idx])) {
-      return string_index_pair{nullptr, 0};
-    }
-
-    auto const d_str  = d_strings.element<string_view>(idx);
-    auto const nchars = d_str.length();
-    int32_t spos      = 0;
-    auto epos         = static_cast<int32_t>(nchars);
-
-    size_type column_count = 0;
-    while (spos <= nchars) {
-      if (prog.find<stack_size>(idx, d_str, spos, epos) <= 0) break;  // no more matches found
-      if (column_count == column_index) break;                        // found our column
-      spos = epos > spos ? epos : spos + 1;
-      epos = static_cast<int32_t>(nchars);
-      ++column_count;
+    auto const match_count = d_counts[idx];
+
```

Comment:
```suggestion
```

Reply:
This is an intentional separation between lines 64 and 66. I prefer this for separation since clang-format will line up the `=` signs if this line is removed and one of the variables is `const` and the other is not. This is a personal preference applying to these lines of code in this particular place.